src/stakingrewards/stakingrewards.js

```diff
@@ -0,0 +1,436 @@
+  epochs = epochs.filter((epoch) => {
+    return epoch.stakes.length > 0
+  })
+
+  // Sort the epoch's stakes by staking provider
+  const stakeList = {}
+  epochs.forEach((epoch) => {
+    epoch.stakes.forEach((epochStake) => {
+      const stakeData = {}
+      stakeData.epochTotalAmount = epoch.totalAmount
+      stakeData.epochDuration = epoch.duration
+      stakeData.epochTimestamp = epoch.timestamp
+      stakeData.amount = epochStake.amount
+      stakeData.epochId = epoch.id
+      stakeList[epochStake.stakingProvider] = stakeList[
+        epochStake.stakingProvider
+      ]
+        ? stakeList[epochStake.stakingProvider]
+        : []
+      stakeList[epochStake.stakingProvider].push(stakeData)
+    })
+  })
+
+  // Calculate the reward of each stake
+  // Rewards formula: r = (s_1 * y_t) * t / 365; where y_t is 0.15
+  const rewards = {}
+  Object.keys(stakeList).map((stakingProvider) => {
+    let reward = BigNumber(0)
+
+    const stake = stakeList[stakingProvider]
+
+    // Check if operator is confirmed and when
+    const opConf = opsConfirmed.find((op) => op.id === stakingProvider)
+    const opConfTimestamp = opConf ? opConf.confirmedTimestamp : undefined
+    if (opConfTimestamp) {
+      reward = stake.reduce((total, epochStake) => {
+        let epochReward = BigNumber(0)
+        const stakeAmount = BigNumber(epochStake.amount)
+        const epochTimestamp = parseInt(epochStake.epochTimestamp)
+        let epochDuration = epochStake.epochDuration
+          ? parseInt(epochStake.epochDuration)
+          : currentTime - epochStake.epochTimestamp
+
+        // If the operator was confirmed in the middle of this epoch...
+        if (
+          opConfTimestamp > epochTimestamp &&
+          opConfTimestamp <= epochTimestamp + epochDuration
+        ) {
+          epochDuration = epochTimestamp + epochDuration - opConfTimestamp
+        }
+
+        epochReward = stakeAmount
+          .times(15)
+          .times(epochDuration)
+          .div(SECONDS_IN_YEAR * 100)
+
+        return total.plus(epochReward)
+      }, BigNumber(0))
+    }
+
+    if (!reward.isZero()) {
+      // Find the beneficiary of this reward
+      const stakeDatasItem = stakeDatas.find(
+        (stake) => stake.id === stakingProvider
+      )
+      const beneficiary = stakeDatasItem.beneficiary
+
+      const stProvCheckSum = ethers.utils.getAddress(stakingProvider)
+      rewards[stProvCheckSum] = {
+        beneficiary: ethers.utils.getAddress(beneficiary),
+        amount: reward.toFixed(0),
+      }
+    }
+  })
+
+  return rewards
+}
+
+/**
+ * Generate the bonus rewards earned by stakes between June 1st and July 15th
+ * and return it in Merkle distribution input format
+ * @param {string}  gqlURL          Subgraph GraphQL API URL
+ * @return {BigNumber}              The amount of generated rewards
+ */
+exports.getBonusMerkleInput = async function (gqlUrl) {
+  const startTimestamp = 1654041600 // Jun 1st 2022 00:00:00 GMT
```

Comment:
Noticed that we hardcde these dates in a couple places. How easily can this package be repurposed if we have to run it a subsequent time with different date ranges? Is it helpful to generalize/parameterize this in some way? Just thinking out loud.

Reply:
It's a good question. The idea here is that the calculation of bonus rewards is something that has happened only once and won't be repeated in the future. So I thought it makes sense that the `start` and `end` timestamps are hardcoded.

Note that this is different in `getOngoingMerkleInput` function. Since this function probably will be reutilised, the timestamps are passed by arguments:

https://github.com/threshold-network/merkle-distribution/blob/08e08e2ee2fe959427003b1fe70ff6e18c9867a7/src/stakingrewards/stakingrewards.js#L235-L239